onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,622 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/propagate_cast_ops.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+std::unordered_set<std::string> fp16_allow = {"Transpose", "Reshape", "Gather", "Split", "Relu", "Where", "Dropout"};
+std::unordered_set<std::string> fp16_safe = { "LayerNorm", "Gelu", "FastGelu", "Tanh", "MatMul", "MatAdd", "Add",
+                                              "Sub", "Mul", "Div", "Neg", "Gemm", "FusedMatMul", "FusedGemm"};
+
```

Comment:
This is a very impressive graph transformer!! Greate job! 
For the ease of maintenance and readability, let's add a few utility functions... I think some could be quite useful for the following functions. 

bool IsCast(Node& node, uint_8 type) , so that we can use it like IsCast(node, k_fp16 | kfp32) to check ... 
bool IsType(NodeArg& node_arg, uint_8 type) to check the nodearg type